### PR TITLE
feat(eslint-plugin): add new rule no-empty-function

### LINKED ISSUE #426: Extend no-empty-function to handle constructors with parameter properties
Spawning off of #48 for clarity.

# What code were you trying to parse?
```TS
export default class Foo {
  constructor(private name: string) {}
}
```

# What happened?
```
  2:36  error  Unexpected empty constructor  no-empty-function
```

### PR DESCRIPTION
**Pull request by scottohara**
Extends the no-empty-function rule from ESlint core to allow for empty constructor bodies when parameter properties are used (a.k.a. concise constructor functions).

Fixes #426

### Examples of allowed code
```typescript
/*eslint @typescript-eslint/no-empty-function: "error"*/

class Person {
  constructor(private firstName: string, private surname: string) {}
}
```

...which is functionally equivalent to writing:

```typescript
class Person {
  private firstName: string;
  private surname: string;

  constructor(firstName: string, surname: string) {
    this.firstName = firstName;
    this.surname = surname;
  }
}
```

Aside from this one specific case, the rule passes through to ESLint core `no-empty-function` for everything else, so the following still applies:

### Examples of valid code

```typescript
/*eslint @typescript-eslint/no-empty-function: ["error", { allow: ['constructors'] }]*/

class Person {
  constructor(firstName: string, surname: string) {}
}
```

```typescript
/*eslint @typescript-eslint/no-empty-function: ["error", { allow: ['methods'] }]*/

class Person {
  otherMethod(firstName: string, surname: string) {}
}
```

### Examples of invalid code

```typescript
/*eslint @typescript-eslint/no-empty-function: "error"*/

class Person {
  constructor(firstName: string, surname: string) {}
}
```

```typescript
/*eslint @typescript-eslint/no-empty-function: "error"*/

class Person {
  otherMethod(firstName: string, surname: string) {}
}
```

### CODE REVIEW

File: packages/eslint-plugin/src/rules/no-empty-function.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      return (
+        node.params &&
+        node.params.some(
+          param => param.type === AST_NODE_TYPES.TSParameterProperty,
+        )
+      );
+    }
+
+    /**
+     *
```
Comment (JamesHenry) [inline packages/eslint-plugin/src/rules/no-empty-function.ts]: Empty JSDoc

**Comment by JamesHenry** [comment]
Definitely one of those awkward ones that is not super clear-cut, but overall seems like a reasonable change and I think matches user expectations

File: packages/eslint-plugin/src/rules/no-empty-function.ts (hunk 1 above)
Comment (scottohara) [reply packages/eslint-plugin/src/rules/no-empty-function.ts]: Fixed

**Comment by bradzacher** [comment]
Code LGTM! Thanks for doing this!

**Comment by joaovieira** [comment]
Just merged an older eslint ruleset (used in JS projects) with a `plugin:@typescript-eslint/recommended` one and started getting `no-useless-constructor` and `no-empty-function` errors as they were enabled in the old ruleset. Just seen both have TS rules now, which is great! ❤️ 

Do you think these could be added to the `recommended` ruleset?

**Comment by bradzacher** [comment]
We only update `recommended` rulesets on breaking changes.
We will discuss the changes to the set when it comes time for a breaking change.

**Comment by onichandame** [comment]
Thx for the fix. But how can I use this PR? Any rules that I can use to override the recommended ruleset?

**Comment by bradzacher** [comment]
please read the rule docs:
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-empty-function.md

**Pull request merged**